### PR TITLE
Dataviews: Add some client side data handling utils

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Follow up of: https://github.com/WordPress/gutenberg/pull/57333

When data handling is done client side there is some code that can be abstracted and reused. In this PR I introduce two of them:
1. Get the pagination info and the paginated results based on the current view
2. Sort text items.

## Testing Instructions
1. Enable the admin views experiment
4. Sorting and pagination(both pagination info like the number of pages etc.. and the paginated results) should work as before.

### CODE REVIEW

File: packages/dataviews/src/utils.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * @param {Object}   params            Function params.
+ * @param {Object[]} params.items      Array of items to sort.
+ * @param {Object}   params.view       Current view object.
+ * @param {Object[]} params.fields     Array of available fields.
+ * @param {string[]} params.textFields Array of the field ids to sort.
+ *
+ * @return {Object[]} Sorted items.
+ */
+export const sortByTextFields = ( { items, view, fields, textFields } ) => {
```
Comment (youknowriad) [inline packages/dataviews/src/utils.js]: I don't understand why we need the text fields argument, isn't the sorting config part of the "view" object

File: packages/dataviews/src/utils.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
+/**
+ * Helper util to sort items by text fields, when sorting is done client side.
+ *
+ * @param {Object}   params            Function params.
+ * @param {Object[]} params.items      Array of items to sort.
```
Comment (youknowriad) [inline packages/dataviews/src/utils.js]: items is called "data" in the dataviews component, should we keep the same name?

File: packages/dataviews/src/utils.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * @param {Object}   params.view  Current view object.
+ *
+ * @return {Object} Paginated items and paginationInfo.
+ */
+export function getPaginationResults( { items, view } ) {
+	const start = ( view.page - 1 ) * view.perPage;
+	const totalItems = items?.length || 0;
+	items = items?.slice( start, start + view.perPage );
+	return {
+		items,
```
Comment (youknowriad) [inline packages/dataviews/src/utils.js]: Should this be "data" as well?

File: packages/dataviews/src/utils.js (hunk 1 above)
Comment (ntsekouras) [reply packages/dataviews/src/utils.js]: For me it seems the type of sorting is related to the field, but if we have an API prop in fields it will only be used for client side(that's what Tanstack does too - sortingFn). So I don't think we need to add such an API like that for now.. Did you have something different in your mind?

File: packages/dataviews/src/utils.js (hunk 2 above)
Comment (ntsekouras) [reply packages/dataviews/src/utils.js]: I can do that.

File: packages/dataviews/src/utils.js (hunk 1 above)
Comment (youknowriad) [reply packages/dataviews/src/utils.js]: The ideal scenario for me is that the component end up with a single hook abstracting all of this at some point:

```
// probably a bad name
const { isLoading, data, paginationInfo } = useLocalRequest( allData, view, fields )
```

separating textFields means that we have to pass an extra argument to such hooks. Isn't the "type" of the field sufficient here?

File: packages/edit-site/src/components/page-patterns/dataviews-patterns.js [hunk 4]
```diff
@@ -229,7 +241,12 @@ export default function DataviewsPatterns() {
 		],
 		[ view.type, categoryId ]
 	);
-
+	// Reset the page number when the category changes.
+	useEffect( () => {
```
Comment (ntsekouras) [inline packages/edit-site/src/components/page-patterns/dataviews-patterns.js:245]: This is an unrelated fix for a bug that I observed here.

File: packages/edit-site/src/components/page-patterns/dataviews-patterns.js [hunk 5]
```diff
@@ -229,7 +241,12 @@ export default function DataviewsPatterns() {
...
+		if ( previousCategoryId !== categoryId && view.page !== 1 ) {
+			setView( { ...view, page: 1 } );
```
Comment (youknowriad) [inline packages/edit-site/src/components/page-patterns/dataviews-patterns.js]: Right now, these categories are built as separate pages, so for me we should reset the entire "view" object when changing categories.

File: packages/edit-site/src/components/page-templates/index.js [hunk 6]
```diff
@@ -252,10 +256,10 @@ export default function DataviewsTemplates() {
 		[ authors, view.type ]
 	);
 
-	const { shownTemplates, paginationInfo } = useMemo( () => {
+	const { data, paginationInfo } = useMemo( () => {
```
Comment (youknowriad) [inline packages/edit-site/src/components/page-templates/index.js:259]: Can this hook become a reusable hook that filters and sorts and everything based on "view", "allData" and "fields"?

File: packages/dataviews/src/utils.js (hunk 1 above)
Comment (ntsekouras) [reply packages/dataviews/src/utils.js]: Hm.. I'm not sure yet if we could use just the `type`. For example in templates `author` the type is `enumeration` and is used for filters and it could be used for inline editing. But for sorting we just want to do it as a `text` field. 

I think things will become clearer when we advance the other APIs to use the `type` in more complex scenarios too.

File: packages/edit-site/src/components/page-patterns/dataviews-patterns.js (hunk 5 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js]: I kind of disagree here because while what you say it's true, it feels like we have external filters in the sidebar and for example if I have sorted by title, I'd expect the other categories to be sorted too. No strong opinions for now though, I can update.

Comment (youknowriad) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js]: No strong opinion too. I don't like this effect though and I'm not sure how to avoid it.

File: packages/dataviews/src/utils.js (hunk 1 above)
Comment (youknowriad) [reply packages/dataviews/src/utils.js]: isn't "enumeration" a special text field?

File: packages/edit-site/src/components/page-templates/index.js (hunk 6 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/page-templates/index.js:259]: We can explore this separately better because I think we need to advance the filters API more and see how we can abstract it best. Right now there is a lot of specific code regarding the filtering(see author in templates).

File: packages/edit-site/src/components/page-patterns/dataviews-patterns.js (hunk 5 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js]: In my mind this is just a filter that we will need to incorporate in the `view` object. Current patterns page is WIP and has many more things to think about - refactor/simplify etc.. It's just that is there is a lot of complexity and want to tackle things iteratively.

File: packages/edit-site/src/components/page-templates/index.js (hunk 6 above)
Comment (youknowriad) [reply packages/edit-site/src/components/page-templates/index.js:259]: It doesn't seem specific at all for me. `author_text` is just `getValue`  and the rest is standard. I'm fine if it's explored separately but I think having something like this is going to help us make the right decisions. And we won't need to expose multiple helpers from the dataviews package only a single hook.

**Review by youknowriad — APPROVED** [review]

File: packages/edit-site/src/components/page-patterns/dataviews-patterns.js (hunk 5 above)
Comment (youknowriad) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js]: This is also close to the "default views" of the pages data views, it's not really clear whether it's filters or just separate default views.

File: packages/dataviews/src/utils.js (hunk 1 above)
Comment (ntsekouras) [reply packages/dataviews/src/utils.js]: Maybe yes, but I'm not too eager to abstract right now because as I said, I think we will be able to make more informed decisions when we evolve the other APIs too.. While not pretty right now, we just pass an array and will be much simpler to change in the near future.

File: packages/edit-site/src/components/page-templates/index.js (hunk 6 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/page-templates/index.js:259]: Of course, eventually we shouldn't have much helpers..

File: packages/edit-site/src/components/page-patterns/dataviews-patterns.js (hunk 5 above)
Comment (oandregal) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js]: > This is also close to the "default views" of the pages data views, it's not really clear whether it's filters or just separate default views.

The "default views" are just presets, preconfigured settings for the view, so we still need a "category" filter in place.
